cmd/kubeadm/app/cmd/phases/workflow/runner.go

```diff
@@ -352,7 +352,7 @@ func (e *Runner) BindToCommand(cmd *cobra.Command) {
 					os.Exit(1)
 				}
 			},
-			Args: cobra.NoArgs, // this forces cobra to fail if a wrong phase name is passed
+			Args: cobra.NoArgs,
```

Comment:
I'm not sure this is correct.
IMO noargs should be applied only for no-leaf phase (see 342)
For leaf phases, I see two options: inherit args function from top level command or allow to specify per phase args function

Reply:
Thinking about this again, IMo there are join phases that actually needs the master art, so this should be changed.

Reply:
This is basically avoiding `kubeadm init/join phase subphase some-arg` when `some-arg` will not be used for anything.

I think it would be benefitial if we instead of accepting everything and ignore it we explicitly fail if more arguments than needed are passed. The big majority of unit phases have no positional arguments. In the case of `init` I don't think there was any, and in `join` we have `discovery/token/token.go` and `preflight/checks.go` that read `cfg.Discovery.BootstrapToken.APIServerEndpoint`, that is what gets overriden by the data initializer if provided as a positional arg.

I think it would be good if we could explicitly set positional argument restrictions on unit phase basis (similar to `InheritFlags`).

Adding more reasoning as a comment outside this conversation.